tests/unit/package_managers/yarn/test_locators.py

```diff
@@ -0,0 +1,359 @@
+    (
+        _ParsedLocator(
+            scope=None,
+            name="npm-lifecycle-scripts",
+            raw_reference="https://github.com/chmeliik/js-lifecycle-scripts.git#workspace=my-workspace&commit=0e786c88d5aca79a68428dadaed4b096bf2ae3e0",
+        ),
+        _ParsedReference(
+            protocol="https:",
+            source="//github.com/chmeliik/js-lifecycle-scripts.git",
+            selector="workspace=my-workspace&commit=0e786c88d5aca79a68428dadaed4b096bf2ae3e0",
+            params=None,
+        ),
+    ),
+    (
+        _ParsedLocator(
+            scope=None,
+            name="ccto-wo-deps",
+            raw_reference="patch:ccto-wo-deps@git@github.com%3Acachito-testing/cachito-npm-without-deps.git%23commit=2f0ce1d7b1f8b35572d919428b965285a69583f6#./.yarn/patches/ccto-wo-deps-git@github.com-e0fce8c89c.patch::version=1.0.0&hash=51a91f&locator=berryscary%40workspace%3A.",
+        ),
+        _ParsedReference(
+            protocol="patch:",
+            source="ccto-wo-deps@git@github.com:cachito-testing/cachito-npm-without-deps.git#commit=2f0ce1d7b1f8b35572d919428b965285a69583f6",
+            selector="./.yarn/patches/ccto-wo-deps-git@github.com-e0fce8c89c.patch",
+            params={
+                "version": ["1.0.0"],
+                "hash": ["51a91f"],
+                "locator": ["berryscary@workspace:."],
+            },
+        ),
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "locator_str, expect_parsed_locator, expect_parsed_reference",
+    [
+        (locator_str, parsed_locator, parsed_reference)
+        for locator_str, (parsed_locator, parsed_reference) in zip(
+            ALL_LOCATORS, PARSED_LOCATORS_AND_REFERENCES
+        )
+    ],
+)
+def test_parse_locator_helper(
+    locator_str: str,
+    expect_parsed_locator: _ParsedLocator,
+    expect_parsed_reference: _ParsedReference,
+) -> None:
+    """Test the helpers that parse locators and references generically.
+
+    Note that these don't care which locator types are supported and which aren't.
+    """
+    locator = _parse_locator(locator_str)
+    assert locator == expect_parsed_locator
+    assert locator.parsed_reference == expect_parsed_reference
+    assert str(locator) == locator_str
+
+
+@pytest.mark.parametrize(
+    "locator_str",
+    [
+        "name",
+        "@scope/name",
+        "name@",
+        "@scope/name@",
+        "@reference",
+        "@scope@reference",
+        "@scope/@reference",
+        "name/@reference",
+    ],
+)
+def test_unexpected_locator_format(locator_str: str) -> None:
+    with pytest.raises(UnexpectedFormat, match="could not parse locator"):
+        _parse_locator(locator_str)
+
+
+def test_unexpected_reference_format() -> None:
+    with pytest.raises(UnexpectedFormat, match="could not parse reference"):
+        # it is very difficult to find something that doesn't match, probably not doable without \n
+        # kudos to hypothesis: https://hypothesis.readthedocs.io/en/latest/index.html
```

Comment:
Wow :O

Reply:
It's pretty cool yeah :smile: 

It can even do stuff like this:
```shell
hypothesis write cachi2.core.package_managers.yarn.locators.parse_locator 
```

```python
# This test code was written by the `hypothesis.extra.ghostwriter` module
# and is provided under the Creative Commons Zero public domain dedication.

import cachi2.core.package_managers.yarn.locators
from hypothesis import given, strategies as st


@given(reference_str=st.text())
def test_fuzz__parse_reference(reference_str: str) -> None:
    cachi2.core.package_managers.yarn.locators._parse_reference(
        reference_str=reference_str
    )
```

But this didn't find the falsifying example, I played with it a bit and arrived at something like this (generate a string that matches the reference regex, double it, try to parse it)

```python
@given(
    reference_str=st.from_regex(
        r"^([^#:]*:)?((?:(?!::)[^#])*)(?:#((?:(?!::).)*))?(?:::(.*))?$"
    )
)
def test_fuzz__parse_reference(reference_str: str) -> None:
    _parse_reference(reference_str=reference_str + reference_str)
```